src/mrpro/algorithms/_cg.py

```diff
@@ -0,0 +1,124 @@
+    problems.
+
+    Also, note that if the condition of H is very large, a small residual does not necessarily
+    imply that the solution is accurate.
+
+    Parameters
+    ----------
+    operator
+        self-adjoint operator (named H above)
+    right_hand_side
+        right-hand-side of the system (named b above)
+    starting_value, optional
+        starting value of the iteration; if None, it will be set to b
+    max_iterations, optional
+        maximal number of iterations
+    tolerance, optional
+        tolerance for the residual
+    callback, optional
+        user-provided function to be called at each iteration
+
+    Returns
+    -------
+        an approximate solution of the linear system Hx=b
+    """
+
+    # throw an error if the shapes of the starting value and the right-hand side
+    # do not coincide
+    if starting_value is not None:
+        if starting_value.shape != right_hand_side.shape:
+            raise ValueError(f'\nshapes {starting_value.shape, right_hand_side.shape} are incompatible')
+
+    # initial residual
+    residual = right_hand_side - operator(starting_value)[0] if starting_value is not None else right_hand_side.clone()
+
+    # initialize conjugate vector
+    conjugate_vector = residual.clone()
+
+    # assign starting value to the solution
+    solution = starting_value.clone() if starting_value is not None else right_hand_side.clone()
+
+    # for the case where the residual is exactly zero
+    if torch.vdot(residual.flatten(), residual.flatten()) == 0:
+        return solution
+
+    # squared tolerance;
+    # (we will check ||residual||^2 < tolerance^2 instead of ||residual|| < tol
+    # #to avoid the computation of the root for the norm)
+    tolerance_squared = tolerance**2
+
+    # dummy value the old squared norm of the residual;
+    # #only required for initialization
+    norm_squared_residual = None
+
+    for i in range(max_iterations):
+
+        # calculate the square norm of the residual
+        residual_flat = residual.flatten()
+        norm_squared_residual_new = torch.vdot(residual_flat, residual_flat).real
+
+        # check if the solution is already accurate enough
+        if tolerance != 0:
+            if norm_squared_residual_new < tolerance_squared:
+                return solution
+
+        if i > 0:
+            beta = norm_squared_residual_new / norm_squared_residual
+            conjugate_vector = residual + beta * conjugate_vector
+
+        # update estimates of the solution and the residual
+        (operator_conjugate_vector,) = operator(conjugate_vector)
+        alpha = norm_squared_residual_new / (
+            torch.vdot(conjugate_vector.flatten(), operator_conjugate_vector.flatten())
+        )
+        solution += alpha * conjugate_vector
+        residual -= alpha * operator_conjugate_vector
+
+        norm_squared_residual = norm_squared_residual_new
```

Comment:
residual_squared = residual_squared_new

Reply:
is the suggestion just to leave out the "norm"? or to have `norm_residual_squared` instead of `residual_norm_squared`?

Reply:
both-ish :D  suggestions would be
1. leave out the norm 
(imho implied by residual squared -> for an physicist a squared vector usually means the norm implied by the dot product, vector^H vector.)
2. to put "residual" first, as this is the important part of the name.

Reply:
ok, I think I prefer to keep the word "norm" in then and go for option 2).